docs/src/prototypes/mentions/MentionsEditor.tsx

```diff
@@ -0,0 +1,160 @@
+import * as React from 'react'
+import * as _ from 'lodash'
+import keyboardKey from 'keyboard-key'
+
+import { atMentionItems, AtMentionItem } from './dataMocks'
+import { insertTextAtCursorPosition } from './utils'
+import { PortalAtCursorPosition } from './PortalAtCursorPosition'
+
+export interface MentionsContainerProps {
+  items?: AtMentionItem[]
+  searchQuery?: string
+  open?: boolean
+  onInputKeyDown?: (e: React.KeyboardEvent) => void
+  onOpenChange?: (e: React.SyntheticEvent, data: { open?: boolean }) => void
+  onSearchQueryChange?: (e: React.SyntheticEvent, data: { searchQuery: string }) => void
+  onSelectedChange?: (e: React.SyntheticEvent, data: { searchQuery: string }) => void
+}
+
+interface MentionsEditorState {
+  shouldUpdate: boolean
+  open: boolean
+  searchQuery: string
+  selectedItem: string
+}
+
+interface MentionsEditorAction {
+  type: 'OPEN' | 'UPDATE_AND_CLOSE' | 'RESET_UPDATE_FLAG' | 'SET_SEARCH_QUERY'
+  args?: string
+}
+
+const editorStyle: React.CSSProperties = {
+  backgroundColor: '#eee',
+  borderRadius: '5px',
+  border: '1px dashed grey',
+  padding: '5px',
+  minHeight: '100px',
+  outline: 0,
+}
+
+const stateReducer = (state: MentionsEditorState, action: MentionsEditorAction) => {
+  switch (action.type) {
+    case 'OPEN':
+      return { ...state, open: true }
+    case 'RESET_UPDATE_FLAG':
+      return { ...state, shouldUpdate: false }
+    case 'UPDATE_AND_CLOSE':
+      return {
+        ...state,
+        shouldUpdate: true,
+        open: false,
+        searchQuery: '',
+        selectedItem: action.args,
+      }
+    case 'SET_SEARCH_QUERY':
+      return { ...state, searchQuery: action.args }
+    default:
+      throw new Error(`Action ${action.type} is not supported`)
+  }
+}
+
+const MentionsEditor: React.FunctionComponent<
+  MentionsContainerProps & { render: (props: MentionsContainerProps) => React.ReactNode }
+> = props => {
+  const { render, ...rest } = props
+  const contendEditableRef = React.useRef(null)
+  const [state, dispatch] = React.useReducer(stateReducer, {
+    shouldUpdate: false,
+    open: false,
+    searchQuery: '',
+    selectedItem: '',
+  })
+
+  React.useEffect(
+    () => {
+      if (!state.shouldUpdate) {
+        return
+      }
+
+      tryFocusEditor()
+
+      // after the wrapped component is closed the value of the search query is inserted in the editor at cursor position
+      insertTextAtCursorPosition(state.selectedItem)
+      dispatch({ type: 'RESET_UPDATE_FLAG' })
+    },
+    [state.shouldUpdate],
+  )
+
+  const handleEditorKeyUp = (e: React.KeyboardEvent) => {
+    if (!state.open && e.shiftKey && keyboardKey.getCode(e) === keyboardKey.AtSign) {
+      dispatch({ type: 'OPEN' })
+    }
+  }
+
+  const handleOpenChange = (e: React.SyntheticEvent, { open }: MentionsContainerProps) => {
+    if (!open) {
```

Comment:
`resetStateAndUpdateEditor()` contains opposite condition...

Reply:
not really, `resetStateAndUpdateEditor` is checking current `open` flag from `state` object and this one is looking at what will be the state of `open` flag coming from `onOpenChange` prop

I agree that the check inside `resetStateAndUpdateEditor` is redundant because `onOpenChange` is fired when `open` flag changes (obviously), so we can just call `dispatch` directly, something like:

```typescript
const handleOpenChange = (e: React.SyntheticEvent, { open }: MentionsContainerProps) => {
  if (!open) {
    dispatch({ type: 'UPDATE_AND_CLOSE', args: state.searchQuery })
  }
}
```

is this better?

Reply:
Yep, I vote for the direct call 👍

Reply:
done